Tariscript (#286)
tari-labs/tari-university: _posts/tari/2021-09-13-TariScript_for_dummies.md

He must make it so that the script executes and returns the second key,  his public key, as a result.

When TariScript is validated, the script is run against an _input stack_ provided by Bob. In this case,
Bob must provide a single element, his public key (`P_b`), as the script input.

Before the first instruction is executed, the stack and script look like this:

| Step   | 0                                           |
|:-------|:--------------------------------------------|
| Script | Dup HashBlake256  PushHash(H_b) EqualVerify |
| Stack  | `P_b`                                       |

The script then runs, executing the instructions in order. If any instruction fails, the entire script fails. When the script completes, there must be a single item left on the stack, and that item must be a valid public key.

The first instruction is `Dup`, which takes the top item on the stack and duplicates it:

| Step   | 1 (`Dup`)                               |
|:-------|:----------------------------------------|
| Script | HashBlake256  PushHash(H_b) EqualVerify |

> | Stack  | `P_b P_b`                               |

Author: NP

blocks. It can do this because spent outputs
are not needed to carry out the overall accounting balance on the system. This is Mimblewimble's primary scaling
advantage over, say, Bitcoin.

Of course, when you prune a UTXO, you also prune its script and input data. This means that if there was a huge
chain reorganisation beyond the pruning horizon, termed a _horizon attack_, then non-archival nodes would be unable to
fully validate some transactions.
They would have to find an archival node to replay the complete transcript between the chain split and the pruning horizon.

It is too much of a handicap to disable pruning in Tari,  so the strategy is to mitigate rather than eliminate horizon
attacks:

* Tari has a very long pruning horizon, several thousand blocks, to make the attack more expensive and less likely.
* Ensure that there is always a critical mass of archival nodes running.
* Users can protect their funds from horizon attacks by simply sending funds to themselves using a standard Mimblewimble transaction.

## End credits

As you can see, the road from the initial idea of "Hey, let's incorporate scripting into Mimblewimble" to the final
product was an adventurous one.

> This feature is particularly gratifying because it was a complete team effort, with at least half a dozen contributors making vital additions. Special thanks go to @BlackwolfSA for

Reviewer: ```suggestion
This feature is particularly gratifying because it was a complete team effort, with at least half a dozen contributors making vital additions. Special thanks go to @[BlackwolfSA](https://github.com/swvheerden) for
```

> breaking my first three or four iterations of this concept, @DavidBurkett for reviewing and coming up with the basic

Reviewer: ```suggestion
breaking my first three or four iterations of this concept, @[DavidBurkett](https://github.com/DavidBurkett) for reviewing and coming up with the basic
```

> script key idea and @MikeTheTike, @SimianZa, @Bizzle and @HansieOdendaal for helping refine and polish TariScript into its

Reviewer: ```suggestion
script key idea and @[MikeTheTike](https://github.com/mikethetike), @[SimianZa](https://github.com/philipr-za), @[Bizzle](https://github.com/delta1) and @[HansieOdendaal](https://github.com/hansieodendaal) for helping refine and polish TariScript into its
```

{: .note .warning }
Math ahead! This section also builds on the concepts we discussed in [Mimblewimble - all the bits](??) and
[Mimblewimble transactions](/protocols/mimblewimble-transactions-explained).

## The TariScript language

TariScript is inspired by and is based on Bitcoin script. Like its older cousin, TariScript is a
[stack-based language](https://en.wikipedia.org/wiki/Stack-oriented_programming). It has strict size limits,
and the instruction set is minimal.

This is for a good reason.

If you've followed this space at all, you're no doubt jaded by the endless stream of headlines marking the various
[hacks](https://therecord.media/hacker-steals-600-million-from-poly-network-in-biggest-cryptocurrency-hack-ever/),
[leaks](https://www.gemini.com/cryptopedia/the-dao-hack-makerdao), and
[oopsies](https://github.com/openethereum/parity-ethereum/issues/6995) stemming at least partly from the fact that smart contracting
platforms like Ethereum and Polygon offer [Turing-complete](https://en.wikipedia.org/wiki/Turing_completeness)
scripting functionality.

tl;dr The more you let users do with your programmable money, the more bugs and vulnerabilities they will expose.

> TariScript, on the other hand, tries to provide as much rope as we need to enable all of teh thingz, without letting you

Reviewer 2: ```suggestion
TariScript, on the other hand, tries to provide as much rope as we need to enable all of the things, without letting you
```

hang yourself. Specifically, we want to keep TariScript small enough that it is amenable to
[formal verification](https://en.wikipedia.org/wiki/Formal_verification). At the very least, we're leaving things open
for TariScript scripts to be formally specified using methodology akin to [Miniscript](http://bitcoin.sipa.be/miniscript/).

This also supports Tari's philosophy of keeping the base layer compact while pushing large-scale contracts and digital
assets platforms onto Layer 2. This is how Tari intends to scale to millions of users and digital assets without having
the entire network come to a grinding halt every time a kitty breeds or an ape gets bored.

> ### An example

Reviewer 2: I think we should describe why it is important to end up with a _specific_ public key after script execution.  Whenever I read the TariScript RFCs, I always seem to wonder why any public key will not be accepted until I find the reason again. This explanation in RFC201 is relevant:
```
The script public key \\(K\_{Si}\\) needed for the script signature verification is not stored with 
the TransactionInput, but obtained by executing the script with the provided input data. 
Because this signature is signed with the script private key \\(k\_{Si}\\), it ensures that only the 
owner can provide the input data \\(\input_i\\) to the TransactionInput.
```

{: .note .info }
For a complete list of Opcodes available in TariScript, see [RFC-0202](https://rfc.tari.com/RFC-0202_TariScriptOpcodes.html).

Let's dive in with the one-sided payments example. The approach to one-sided payments is that Alice sends Tari to Bob
using a shared secret in the UTXO commitment and a script locked to Bob's private key. She does this without Bob
getting involved at any point.

This means that both Alice and Bob have a key to unlock the first lock in the door. This situation is not ideal for Bob.

But Bob is mollified since only he has the key to the second lock, so ultimately, only he can spend these funds.

Let's take a closer look at how the script resolution works.

When Alice creates the transaction, she looks up Bob's public key (`P_b`) and hashes it to `H_b`.
She attaches the following script to the UTXO:

```
  Dup HashBlake256  PushHash(H_b) EqualVerify
```

Alice signs the script along with some other housekeeping chores and broadcasts the transaction.

Later, when Bob comes online, his wallet scans the blockchain for UTXOs that have the script matching his public key

> hash.

Reviewer 2: It may be worthwhile to mention that Bulletproof rewinding is used to obtain the commitment value after the first (Mimblewimble) lock is obtained, which is also needed for script signature verification.
- scans blockchain for known one-sided scripts trying to unlock the script with Diffie-Hellmann key exchange

> Now to spend this UTXO, Bob needs the keys for both locks. He gets the key for the first (Mimblewimble) lock using [Diffie-Hellmann key exchange](https://en.wikipedia.org/wiki/Diffie–Hellman_key_exchange) with Alice's public data in the transaction.

Reviewer 2: It may be worthwhile to mention that Bulletproof rewinding is used to obtain the commitment value after the first (Mimblewimble) lock is obtained, which is also needed for script signature verification:
- scans blockchain for known one-sided scripts matching his public key
- obtain the first (Mimblewimble) lock with Diffie-Hellmann key exchange
- obtain the value with Bulletproof rewinding using the first (Mimblewimble) lock

|:-------|:----------------------------------------|
| Script | HashBlake256  PushHash(H_b) EqualVerify |
| Stack  | `P_b P_b`                               |

The next instruction, `HashBlake256` hashes the top stack item and pushes the result onto the stack. We'll keep the same nomenclature that Alice used and note that the hash of `P_b` is `H_b`:

| Step   | 2 (`HashBlake256`)        |
|:-------|:--------------------------|
| Script | PushHash(H_b) EqualVerify |
| Stack  | `H_b P_b`                 |

Next, we have `PushHash(H_b)`. As you might guess, this pushes the value `H_b` onto the stack:

| Step   | 3(`PushHash(H_b)`) |
|:-------|:-------------------|
| Script | EqualVerify        |
| Stack  | `H_b H_b P_b`      |

Finally, the last instruction, `EqualVerify`, removes and compares the top two stack items. If they're equal, the script execution continues. If not, the script fails.

Assuming the hashes match, i.e. Bob provided a public key that matched the hash Alice provided, then the script completes
successfully leaving Bob's public key on the stack.

| Step   | 4 - Complete |
|:-------|:-------------|
| Script |              |
| Stack  | `P_b`        |

> The rules of TariScript say that the key for the second lock is the private key corresponding to `P_b`, which only Bob

Reviewer 3: ```suggestion
The rules of TariScript say that the key for the second lock is the private key of the public key left on the stack after the execution of the script, in this example `p_b` corresponding to `P_b`, which only Bob
```
This concept is crucial to understanding how Tari script works but is only implied. Its never stated

### The metadata signature

Before TariScript, UTXO output features could be manipulated by miners without detection. The metadata
signature closes that vulnerability and locks down all the data except for the range proof, which is already non-malleable.

The metadata signature is a joint, or aggregated, signature that is signed by the _commitment_ (i.e. the combination of the
spend key and value), which Bob the receiver usually knows, and the _spender offset key_, which Alice usually knows.

Notice that Bob signs with the commitment and not just the spend public key. If we did use just the spend public key,
we would leak the value of most transactions.

{: .note .info }
The reason we would leak the values of transactions is a follows: Because $ C = k.G + v.H = P + V $, if we know $ P $,
then $ V $ is trivial to calculate. Now with $ V = v.H $, we can pre-calculate a long list of $V$ for the most commonly
used $v$ values (10, 1000, 10,000, 20,000 etc.) and have a simple lookup table to break confidentiality in the majority
of Tari transactions.

> As a result, we must use a twist on the Schnorr signature called a commitment signature. The details of commitment signatures are very straightforward and make an excellent topic for another TLU post.

Reviewer 3: ```suggestion
As a result, we must use a twist on the Schnorr signature called a commitment signature. The details of commitment signatures are very straightforward and make an excellent topic for another TLU post, but the tl;dr is that it's a combination of two Schnorr signatures on two different generator points. 
```
I think we need a tiny bit more here as well.

### Replay attacks

The sender offset key also prevents _replay attacks_.

In the naïve implementation of TariScript, consider the following scenario:

1. Alice sends a one-sided payment to Bob (Alice has the key for the first lock, because the commitment spend key is a shared secret).
2. Bob later spends the UTXO by sending funds to Charlie. Bob has the key for the second lock, and presents the proof of such when constructing the transaction.
3. Later, Alice sends another one-sided payment to Bob. Again, Alice has a key for the first lock.
4. She then conspires with Charlie, to _reuse_ the signature that Bob used when sending funds to Charlie. Even though they don't have the key to the first
   lock, the input signature is sufficient to convince everyone that they do know it.

There are some steps one could take to make this attack more difficult, such as adding the commitment to the signature
challenge. This still would not eliminate the attack, since Alice would then be careful to use the same shared secret
and value (and thus the same commitment) in her second transaction to Bob.

> However, the sender offset key _does_ block the replay attack.

Reviewer 3: This is not true. 
Charlie has to clone the entire UTXO he received again, with this he is cloning the entire script offset as well.  
We currently block replay attacks by disallowing duplicate commitments, so neither Alice or Charlie can recreate the duplicate UTXO's to do this.

Reviewer 3: See: https://github.com/tari-project/tari/blob/development/docs/src/replay_attacks.md

> Alice and Bob cannot construct the replay transaction without knowledge of the second script offset, which only Bob

Reviewer 3: ```suggestion
Alice and Bob cannot construct the replay transaction without knowledge of the second script offset, which only Bob
```
```suggestion
Alice and Charlie cannot construct the replay transaction without knowledge of the second script offset, which only Bob
```
